CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+
+
+--Panel update methods
+function Panel:SetPartCFrame(cframe)
+	if not self.part then
+		return
+	end
+	self.part.CFrame = cframe * CFrame.new(0, 0, -0.5)
+end
+
+function Panel:SetEnabled(enabled)
+	if self.isEnabled == enabled then
+		return
+	end
+
+	self.isEnabled = enabled
+	if not enabled and (not self.part or not self.gui) then
+		return
+	else
+		self:GetPart()
+		self:GetGUI()
+	end
+	if enabled then
+		self.part.Parent = workspace.CurrentCamera --todo: Perhaps this can change soon.
+		self.gui.Enabled = true
+		for i, v in pairs(self.subpanels) do
+			v.part.Parent = workspace.CurrentCamera
+			v.gui.Enabled = true
+		end
+	else
+		self.part.Parent = nil
+		self.gui.Enabled = false
+		for i, v in pairs(self.subpanels) do
+			v.part.Parent = nil
+			v.gui.Enabled = false
+		end
+	end
+
+	self:OnEnabled(enabled)
+end
+
+function Panel:EvaluatePositioning(cameraCF, cameraRenderCF, userHeadCF)
+	if self.panelType == Panel3D.Type.Floor then
+		--Floor panels simply... go on the floor.
+		--Panel will be in camera's local space (which is assumed to be horizontal),
+		--and floorRotation is derived from the user's head rotation (yaw only)
+		local floorCF = cameraCF * CFrame.new(0, -headHeightFromFloor * currentHeadScale, 0) * floorRotation
+		self:SetPartCFrame(floorCF * pointUpCF)
+	elseif self.panelType == Panel3D.Type.Fixed then
+		--Places the panel in the camera's local space, but doesn't follow the user's head.
+		--Useful if you know what you're doing. localCF can be updated in PreUpdate for animation.
+		local cf = self.localCF - self.localCF.p
+		cf = cf + (self.localCF.p * currentHeadScale)
+		self:SetPartCFrame(cameraCF * self.localCF)
+	elseif self.panelType == Panel3D.Type.HorizontalFollow then
+		local headLook = userHeadCF.lookVector
+		local headYaw = math.atan2(-headLook.Z, headLook.X) + math.rad(90)
+		local headForwardCF = CFrame.Angles(0, headYaw, 0) + userHeadCF.p
+		local localCF = (headForwardCF * self.angleFromForward) * --Rotate about Y (left-right)
+						self.angleFromHorizon * --Rotate about X (up-down)
+						(currentHeadScale * self.distance) * --Move into scene
+						turnAroundCF --Turn around to face character
+		self:SetPartCFrame(cameraCF * localCF)
+	elseif self.panelType == Panel3D.Type.FixedToHead then
+		--Places the panel in the user's head local space. localCF can be updated in PreUpdate for animation.
+		local cf = self.localCF - self.localCF.p
+		cf = cf + (self.localCF.p * currentHeadScale)
+		self:SetPartCFrame(cameraRenderCF * cf)
+	end
+end
+
+function Panel:EvaluateGaze(cameraCF, cameraRenderCF, userHeadCF, lookRay)
+	--reset distance data
+	self.isClosest = false
+	self.lookAtPixel = zeroVector2
+	self.lookAtDistance = math.huge
+
+	--Evaluate the lookRay versus this panel
+	local planeCF = self.part.CFrame
+	local planeNormal = planeCF.lookVector
+	local pointOnPlane = planeCF.p + (planeNormal * 0.5) --Move the point out by half the thickness of the part (part is *always* 1 stud thick)
```

Comment:
Is it possible to programmatically extract this 0.5 value, or at the very least define the thickness in a constant so we can re-use that here.